File: diagnostic_aggregator/src/diagnostic_aggregator/generic_analyzer.py
#! /usr/bin/env python
#
# Software License Agreement (BSD License)
#
# Copyright (c) 2008, Willow Garage, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of the Willow Garage nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#

# Author: Kevin Watts

##\brief GenericAnalyzer is a basic analyzer.

PKG = 'diagnostic_aggregator'
import roslib; roslib.load_manifest(PKG)

import rospy
from xml.dom import minidom

from diagnostic_msgs.msg import DiagnosticStatus, KeyValue

stat_dict = { 0: 'OK', 1: 'Warning', 2: 'Error' }

from diagnostic_aggregator.analyzer import *

##\brief Analyzes all data for selected topics. Default analyzer.
##
## GenericAnalyzer is initialized by XML snippet:
##\verbatim
## <analyzer type="GenericAnalyzer" pkg="diagnostic_aggregator" 
##           file="generic_analyzer" prefix="Motors" 
##           expected="my_expected" >
##  <topic startswith="EtherCAT" />
##  <topic startswith="Realtime" />
##  <topic contains="Power Node" />
##  <topic name="Mechanism Control" />
##</analyzer>
##\endverbatim
## The analyzer will group together selected topics according to the options 
## selected by the user. Users can specify status names by 'startswith', 
## 'contains', or 'name' for an exact name. 
## 
## Any expected names can be set with the 'expected' parameter name. The analyzer
## will read the '~my_expected' parameter from the parameter server and compare it
## to the list of topics received. Any topics in the expected list do not need to be
## set with the 'topic' tag.
## 
## Analyzers are dynamically loaded by the aggregator.
## GenericAnalyzer stores, analyzes and outputs [ DiagnosticStatus ]
class GenericAnalyzer:
    ##\param xml_node minidom node : XML node to initialize analyzer
    ##\param prefix str : Prefix to be appended to all output 
    ##\param other bool (optional) : Handles remaining messages if true
    def __init__(self, xml_node, prefix, other = False):
        self._prefix = prefix

        self._status_names = {} # Store by last time

        self._other = other
        if self._other:
            return
        
        self._startswith = []
        self._names = []
        self._contains = []

        self._expected = []

        # Any names that are expected are automatically added to the 
        # names list, with error state and 'Missing' message
        if xml_node.attributes.has_key('expected'):
            expected_param = xml_node.attributes['expected'].value
            params = rospy.get_param('~%s' % expected_param, None)
            if not params:
                raise ParameterNotFoundError
            for name in params:
                self._expected.append(name)
                exp_stat = DiagnosticStatus()
                exp_stat.name = name
                exp_stat.level = 2
                exp_stat.message = 'Missing'
                self._status_names[name] = DiagnosticItem(exp_stat)

        # Load remaining topic nodes
        for topic_node in xml_node.getElementsByTagName('topic'):
            topic_attrs = topic_node.attributes.keys()
            
            if 'startswith' in topic_attrs:
                starts_with = topic_node.attributes['startswith'].value
                self._startswith.append(starts_with)

            if 'name' in topic_attrs:
                self._names.append(str(topic_node.attributes['name'].value))

            if 'contains' in topic_attrs:
                self._contains.append(str(topic_node.attributes['contains'].value))

        
    ##\brief Analyze messages, returns processed array
    ## 
    ## Analyze messages and returns list of analyzed DiagnosticStatus msgs
    ##\param msgs dict{DiagnosticStatus, count} : count++ if analyzed by this analyzer
    def analyze(self, msgs):
        status_array = []

        header_status = DiagnosticStatus()
        header_status.name = str(self._prefix)
        header_status.level = 0
        header_status.message = 'OK'
        header_status.values = []

        status_array.append(header_status)
        
        # Find messages to analyze
        msgs_to_analyze = self._select_msgs(msgs)
        
        self._update_status_names(msgs_to_analyze)

        # Process messages
        all_stale = True
        for name, item in self._status_names.iteritems():
            # Update stale
            update_time = rospy.get_time() - item._update_time
            stale = update_time > 3.0 ##@todo Make this parameter
            all_stale = stale and all_stale

            # Update level
            header_status.level = max(header_status.level, item._level)
            level_str = stat_dict[item._level]

            # Automatically error header if it's stale
            if stale:
                header_status.level = 2
                level_str = 'Stale'

            # Add to header DiagnosticStatus
            header_status.values.append(KeyValue(name, level_str))

            # Output status message
            status_array.append(item.to_status_msg(self._prefix, stale))

        header_status.message = stat_dict[header_status.level]

        header_status.name = header_status.name 

        if all_stale:
            header_status.level = 3
            header_status.message = 'All Stale'
    
        return status_array

    ##\brief Returns messages to analyze based on topic name
    ##
    ## Checks all status names if they need analysis, updates analysis
    ## count of status names. Checks status.name against expected, startswith,
    ## contains and name lists.
    ##\param msgs dict{DiagnosticStatus, count} : Given array of messages
    def _select_msgs(self, msgs):
        to_analyze = []
        
        # If processing the remainder, return all unanalyzed messages
        if self._other:
            for msg, value in msgs.iteritems():
                if value == 0:
                    to_analyze.append(msg)

            return to_analyze

        for msg, value in msgs.iteritems():
            for expected in self._expected:
                if msg.name == expected:
                    to_analyze.append(msg)
                    msgs[msg] += 1
                    continue

            for startswith in self._startswith:
                if msg.name.startswith(startswith):
                    to_analyze.append(msg)
                    msgs[msg] += 1
                    continue

            for name in self._names:
                if msg.name == name:
                    to_analyze.append(msg)
                    msgs[msg] += 1
                    continue

            for contain in self._contains:
                if msg.name.find(contain) > -1:
                    to_analyze.append(msg)
                    msgs[msg] += 1
                    continue

        return to_analyze

    ##\brief Updates stored array of status messages 
    ##\param msgs list : List of DiagnosticStatus messages
    def _update_status_names(self, msgs):
        for msg in msgs:
            if not self._status_names.has_key(msg.name):
                self._status_names[msg.name] = DiagnosticItem(msg)
            self._status_names[msg.name].update(msg)


        
        

        
                

File: diagnostic_aggregator/src/generic_analyzer.cpp
/*********************************************************************
 * Software License Agreement (BSD License)
 *
 *  Copyright (c) 2009, Willow Garage, Inc.
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *   * Neither the name of the Willow Garage nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *********************************************************************/

/**< \author Kevin Watts */
/**< \author Arne Nordmann */

#include "diagnostic_aggregator/generic_analyzer.hpp"

#include <map>
#include <memory>
#include <string>
#include <vector>

#include "rclcpp/parameter.hpp"

PLUGINLIB_EXPORT_CLASS(diagnostic_aggregator::GenericAnalyzer, diagnostic_aggregator::Analyzer)

namespace diagnostic_aggregator
{
using std::string;
using std::vector;

GenericAnalyzer::GenericAnalyzer() {}

bool GenericAnalyzer::init(
  const std::string & path, const std::string & breadcrumb, const rclcpp::Node::SharedPtr n)
{
  path_ = path;
  breadcrumb_ = breadcrumb;
  nice_name_ = breadcrumb;
  RCLCPP_DEBUG(
    rclcpp::get_logger("GenericAnalyzer"), "GenericAnalyzer, breadcrumb: %s", breadcrumb_.c_str());

  std::map<std::string, rclcpp::Parameter> parameters;
  if (!n->get_parameters(breadcrumb_, parameters)) {
    RCLCPP_ERROR(
      rclcpp::get_logger("GenericAnalyzer"),
      "Couldn't retrieve parameters for generic analyzer at prefix '%s'.", breadcrumb_.c_str());
    return false;
  }
  RCLCPP_DEBUG(
    rclcpp::get_logger("GenericAnalyzer"), "Retrieved %zu parameter(s) for prefix '%s'.",
    parameters.size(), breadcrumb_.c_str());

  double timeout = 5.0;
  int num_items_expected = -1;
  bool discard_stale = false;

  for (const auto & param : parameters) {
    string pname = param.first;
    rclcpp::Parameter pvalue = param.second;

    if (pname.compare("path") == 0) {
      RCLCPP_DEBUG(
        rclcpp::get_logger("GenericAnalyzer"), "GenericAnalyzer '%s' found path: %s",
        nice_name_.c_str(), pvalue.value_to_string().c_str());
      nice_name_ = pvalue.as_string();
    } else if (pname.compare("find_and_remove_prefix") == 0) {
      RCLCPP_DEBUG(
        rclcpp::get_logger("GenericAnalyzer"),
        "GenericAnalyzer '%s' found find_and_remove_prefix: %s", nice_name_.c_str(),
        pvalue.value_to_string().c_str());
      vector<string> output = pvalue.as_string_array();
      chaff_ = output;
      startswith_ = output;
    } else if (pname.compare("remove_prefix") == 0) {
      RCLCPP_DEBUG(
        rclcpp::get_logger("GenericAnalyzer"), "GenericAnalyzer '%s' found remove_prefix: %s",
        nice_name_.c_str(), pvalue.value_to_string().c_str());
      chaff_ = pvalue.as_string_array();
    } else if (pname.compare("startswith") == 0) {
      RCLCPP_DEBUG(
        rclcpp::get_logger("GenericAnalyzer"), "GenericAnalyzer '%s' found startswith: %s",
        nice_name_.c_str(), pvalue.value_to_string().c_str());
      startswith_ = pvalue.as_string_array();
    } else if (pname.compare("contains") == 0) {
      RCLCPP_DEBUG(
        rclcpp::get_logger("GenericAnalyzer"), "GenericAnalyzer '%s' found contains: %s",
        nice_name_.c_str(), pvalue.value_to_string().c_str());
      contains_ = pvalue.as_string_array();
    } else if (pname.compare("expected") == 0) {
      RCLCPP_DEBUG(
        rclcpp::get_logger("GenericAnalyzer"), "GenericAnalyzer '%s' found expected: %s",
        nice_name_.c_str(), pvalue.value_to_string().c_str());
      for (auto exp : pvalue.as_string_array()) {
        auto item = std::make_shared<StatusItem>(exp);
        this->addItem(exp, item);
      }
    } else if (pname.compare("regex") == 0) {
      RCLCPP_DEBUG(
        rclcpp::get_logger("GenericAnalyzer"), "GenericAnalyzer '%s' found regex: %s",
        nice_name_.c_str(), pvalue.value_to_string().c_str());
      for (auto regex : pvalue.as_string_array()) {
        try {
          std::regex re(regex);
          regex_.push_back(re);
        } catch (std::regex_error & e) {
          RCLCPP_ERROR(
            rclcpp::get_logger("GenericAnalyzer"),
            "Attempted to make regex from %s. Caught exception, ignoring value. Exception: %s",
            regex.c_str(), e.what());
        }
      }
    } else if (pname.compare("timeout") == 0) {
      RCLCPP_DEBUG(
        rclcpp::get_logger("GenericAnalyzer"), "GenericAnalyzer '%s' found timeout: %s",
        nice_name_.c_str(), pvalue.value_to_string().c_str());
      timeout = pvalue.as_double();
    } else if (pname.compare("num_items") == 0) {
      RCLCPP_DEBUG(
        rclcpp::get_logger("GenericAnalyzer"), "GenericAnalyzer '%s' found num_items: %s",
        nice_name_.c_str(), pvalue.value_to_string().c_str());
      num_items_expected = static_cast<int>(pvalue.as_int());
    } else if (pname.compare("discard_stale") == 0) {
      RCLCPP_DEBUG(
        rclcpp::get_logger("GenericAnalyzer"), "GenericAnalyzer '%s' found discard_stale: %s",
        nice_name_.c_str(), pvalue.value_to_string().c_str());
      discard_stale = pvalue.as_bool();
    }
  }

  if (
    startswith_.size() == 0 && name_.size() == 0 && contains_.size() == 0 &&
    expected_.size() == 0 && regex_.size() == 0)
  {
    RCLCPP_ERROR(
      rclcpp::get_logger("generic_analyzer"),
      "GenericAnalyzer '%s' was not initialized with any way of checking diagnostics."
      "Name: %s, namespace: %s",
      nice_name_.c_str(), path.c_str(), n->get_namespace());
    return false;
  }

  // convert chaff_ to output name format. Fixes #17
  for (size_t i = 0; i < chaff_.size(); i++) {
    chaff_[i] = getOutputName(chaff_[i]);
  }

  string my_path;
  if (path == "/") {
    my_path = nice_name_;
  } else {
    my_path = path + "/" + nice_name_;
  }

  if (my_path.find("/") != 0) {
    my_path = "/" + my_path;
  }

  return GenericAnalyzerBase::init(path_, breadcrumb_, timeout, num_items_expected, discard_stale);
}

GenericAnalyzer::~GenericAnalyzer() {}

bool GenericAnalyzer::match(const string & name)
{
  RCLCPP_DEBUG(
    rclcpp::get_logger("GenericAnalyzer"), "Analyzer '%s' match %s", nice_name_.c_str(),
    name.c_str());

  std::cmatch what;
  for (unsigned int i = 0; i < regex_.size(); ++i) {
    if (std::regex_match(name.c_str(), what, regex_[i])) {
      RCLCPP_INFO(
        rclcpp::get_logger("GenericAnalyzer"), "Analyzer '%s' matches '%s' with regex.",
        nice_name_.c_str(), name.c_str());
      return true;
    }
  }

  for (unsigned int i = 0; i < expected_.size(); ++i) {
    if (name == expected_[i]) {
      RCLCPP_INFO(
        rclcpp::get_logger("GenericAnalyzer"), "Analyzer '%s' matches '%s'.", nice_name_.c_str(),
        name.c_str());
      return true;
    }
  }

  for (unsigned int i = 0; i < name_.size(); ++i) {
    if (name == name_[i]) {
      RCLCPP_INFO(
        rclcpp::get_logger("GenericAnalyzer"), "Analyzer '%s' matches '%s'.", nice_name_.c_str(),
        name.c_str());
      return true;
    }
  }

  for (unsigned int i = 0; i < startswith_.size(); ++i) {
    if (name.find(startswith_[i]) == 0) {
      RCLCPP_INFO(
        rclcpp::get_logger("GenericAnalyzer"), "Analyzer '%s' matches '%s'.", nice_name_.c_str(),
        name.c_str());
      return true;
    }
  }

  for (unsigned int i = 0; i < contains_.size(); ++i) {
    if (name.find(contains_[i]) != string::npos) {
      RCLCPP_INFO(
        rclcpp::get_logger("GenericAnalyzer"), "Analyzer '%s' matches '%s'.", nice_name_.c_str(),
        name.c_str());
      return true;
    }
  }

  return false;
}

vector<std::shared_ptr<diagnostic_msgs::msg::DiagnosticStatus>> GenericAnalyzer::report()
{
  RCLCPP_DEBUG(rclcpp::get_logger("GenericAnalyzer"), "Analyzer '%s' report()", nice_name_.c_str());

  vector<std::shared_ptr<diagnostic_msgs::msg::DiagnosticStatus>> processed =
    GenericAnalyzerBase::report();

  // Check and make sure our expected names haven't been removed ...
  vector<string> expected_names_missing;
  bool has_name = false;

  for (unsigned int i = 0; i < expected_.size(); ++i) {
    has_name = false;
    for (unsigned int j = 0; j < processed.size(); ++j) {
      size_t last_slash = processed[j]->name.rfind("/");
      string nice_name = processed[j]->name.substr(last_slash + 1);
      if (nice_name == expected_[i] || nice_name == getOutputName(expected_[i])) {
        has_name = true;
        break;
      }

      // Remove chaff, check names
      for (unsigned int k = 0; k < chaff_.size(); ++k) {
        if (nice_name == removeLeadingNameChaff(expected_[i], chaff_[k])) {
          has_name = true;
          break;
        }
      }
    }
    if (!has_name) {
      expected_names_missing.push_back(expected_[i]);
    }
  }

  // Check that all processed items aren't stale
  bool all_stale = true;
  for (unsigned int j = 0; j < processed.size(); ++j) {
    if (processed[j]->level != diagnostic_msgs::msg::DiagnosticStatus::STALE) {
      all_stale = false;
    }
  }

  // Add missing names to header ...
  for (unsigned int i = 0; i < expected_names_missing.size(); ++i) {
    std::shared_ptr<StatusItem> item(new StatusItem(expected_names_missing[i]));
    processed.push_back(item->toStatusMsg(path_, true));
  }

  for (unsigned int j = 0; j < processed.size(); ++j) {
    // Remove all leading name chaff
    for (unsigned int i = 0; i < chaff_.size(); ++i) {
      processed[j]->name = removeLeadingNameChaff(processed[j]->name, chaff_[i]);
    }

    // If we're missing any items, set the header status to error or stale
    if (expected_names_missing.size() > 0 && processed[j]->name == path_) {
      if (!all_stale) {
        processed[j]->level = diagnostic_msgs::msg::DiagnosticStatus::ERROR;
        processed[j]->message = "Error";
      } else {
        processed[j]->level = diagnostic_msgs::msg::DiagnosticStatus::STALE;
        processed[j]->message = "All Stale";
      }

      // Add all missing items to header item
      for (unsigned int k = 0; k < expected_names_missing.size(); ++k) {
        diagnostic_msgs::msg::KeyValue kv;
        kv.key = expected_names_missing[k];
        kv.value = "Missing";
        processed[j]->values.push_back(kv);
      }
    }
  }

  return processed;
}

}  // namespace diagnostic_aggregator
